### ISSUE TITLE
Document file system layout for qTox

### ISSUE THREAD

**Issue by sudden6** [qTox/qTox#5410]
##### Observed Behavior

Currently qTox uses different methods to compute the paths for the `profile.ini` `profile.db` `qTox.ini` `profile.tox` and the themes folder path.

##### Expected Behavior

- qTox will use `QStandardPaths` for it's own directories and implement a class to abstract the paths for files defined in the Tox Client Standard.

- qTox will follow the freedesktop.org specifications on Linux and store data/config files in the appropriate locations

**Comment by sudden6** [comment]
After some offline discussion with @sphaerophoria I propose the following filesystem layout changes for qTox:

- move the global `qtox.ini` to `QStandardPaths::AppConfigLocation`
- move `<profile>.db` and `<profile>.ini` to `QStandardPaths::AppDataLocation`
- default filetransfer directory will be `QStandardPaths::AppDataLocation/transfers/<PubKey>/`

- the `<profile>.tox` file and the avatars will stay at `~/.config/share/tox/` because this location is given by the TCS in https://tox.gitbooks.io/tox-client-standard/content/data_storage/data_storage.html and https://tox.gitbooks.io/tox-client-standard/content/data_storage/export_format.html

on Linux this would result in the following paths:
- `~/.config/qTox/qtox.ini`
- `~/.local/share/qTox/transfers/<PubKey>`
- `~/.local/share/qTox/<profile>.db`
- `~/.local/share/qTox/<profile>.ini`

The main reason for this change is that currently there are some hardcoded paths for Linux that don't respect https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html. Additionally with that change I want to prevent a mess because there are slightly different implementations inside qTox on how to access these paths and centralize it.

**Comment by sudden6** [comment]
@anthonybilinski @Diadlo @sphaerophoria do you agree with that change?

**Comment by sphaerophoria** [comment]
that sounds reasonable, only possible change I would make is put transfers in AppLocalDataLocation. I'm not sure what the normal policy is for storing lots of files in the roaming folder on windows is.

*However* I'm not entirely convinced that the value for moving the files is worth the pain of having to support old locations forever. We could also follow freedesktop spec using tox as our appname instead of qTox just to avoid having to move anything and support the old locations for the rest of time.

**Comment by sudden6** [comment]
> that sounds reasonable, only possible change I would make is put transfers in AppLocalDataLocation. I'm not sure what the normal policy is for storing lots of files in the roaming folder on windows is.

Good point, I'm not sure how this is supposed to work on Windows tbh, but it might be bad to move a lot of data over the network with roaming profiles.

> having to support old locations forever

I think we should support the old locations for 2-3 versions or 1 year, after that I think we should just remove the old code. Would you see that as more acceptable?

**Comment by anthonybilinski** [comment]
I'm fine with the using the new locations conceptually. I think by far the main issue is then how to update or stay compatible with data saved in the old locations. I don't like the idea of keeping upgrade code for a couple revs then removing, because then if a user upgrades after a year or two they'll silently have all their data missing and won't know why.

I think there are two main ways we could do this and still guarantee users won't lose data:
1. Keep upgrade code in code base indefinitely
    1. nice for user, gross for maintenance
2. Version settings, keep upgrade logic for a couple major revs, if we're starting qTox and we don't know how to upgrade the saved settings because they're too old and we've removed the upgrade logic, tell the user to downgrade to version X so that settings can be upgraded, then re-upgrade to the next version, then the next, then back to the one they were trying to.
    1. less nice for users (but still safe and clear even in worst case), should be painfree for most users assuming they update (I'm not sure how true this is in reality?), and would be much nicer for maintenance because we wouldn't have ever growing upgrade code that becomes more and more out of date.

Talking to @sudden6, he'd prefer 2) because of the maintenance pain of infinite backwards compatibility. I think that's fair, I'm fine with that too. I think whatever we choose for upgrade logic we should also follow for DB versioning and upgrade, and should reflect these setting versions in our qTox version number, maybe by bumping the major rev when there's a DB or save data breaking change?

**Comment by sudden6** [comment]
> keep upgrade logic for a couple major revs

To be more precise here, I think about 3 (arbitrarily choosen) non-bugfix releases of qTox. ie. 1.18.0 to 1.18.1 would not count. 1.18.0 to 1.21.0 would be the last supported upgrade path.

> reflect these setting versions in our qTox version number

I think this could have a negative effect on distros updating to more recent versions of qTox, but if I'm wrong here I'd be fine with e.g. the next version being 2.0.0.

**Comment by anthonybilinski** [comment]
After some more discussion with @sphaerophoria and @sudden6, I think having the indefinite backwards compatibility is the better option. The maintenance cost would be very low since the upgrade logic could be in its own class in its own file with almost no interaction with the rest of the project. It could easily be unit tested, and would never need to change after implementation. The user experience would be much better than having to manually downgrade and upgrade, and even for qTox to check if it can upgrade from the version like in 2), it would still need to always remember the old save file locations, which is half of the indefinite upgrade burden anyway.

For overall architecture, I think having version numbers in config files / DB is a good idea, then having each upgrade function handle only a single revision upgrade, then chaining them together when needed to upgrade multiple revisions at once up to the running qTox version is simplest.
e.g.
````
constexpr Version current_version = {1.17.3};
Version GetSaveVersion()
{
    if (old_file_location_present())
         return {1.16.3};
    else
         return get_vers_from_file(new_file_location)
}

void UpgradeData()
{
	Version oldVer = GetSaveVersion();
	switch(oldVer)
	{
		case Version{1.16.3}:
			UpgradeFrom1_16_3();
			// fallthrough
		case Version{1.17.0}:
			UpgradeFrom1_17_0();
			// fallthrough
		case Version{1.17.1}:
			UpgradeFrom1_17_1();
			// fallthrough
		case Version{1.17.2}:
			UpgradeFrom1_17_2();
			qInfo() << "Updated save files from" << oldVer << "to" << current_version;
			break;
		case Version{1.17.3}:
			qDebug() << "Save files are up to date at version" << current_version;
			break;
		default:
			qCritical() << "Unknown save files ahhhhhh";

	}
}

void UpgradeFrom1_16_3()
{
	// move files to XDG spec locations, set version numbers
}
void UpgradeFrom1_17_0()
{
	// remove some settings, bump version numbers in files
}
void UpgradeFrom1_17_1()
{
	// rename some settings, bump version numbers in files
}
void UpgradeFrom1_17_2()
{
	// update DB layout or something, bump version numbers in files
}
````

**Comment by sudden6** [comment]
Would only `qtox.ini` be versioned or also `profile.ini`?
